### PR TITLE
feat(#1): implement buckets per-repository infrastructure

### LINKED ISSUE #1: Implement per-repository S3 state buckets for test & prod AWS environments
### Description
Implement infrastructure that provisions a dedicated S3 state bucket for each repository, following AWS best practices.
For every repository, the system must automatically create:

An S3 bucket named:
pulumi-<repo>-state
Versioning enabled
Lifecycle rules to expire non-current versions
Default SSE-S3 encryption
Full public access block
Bucket policy enforcing TLS-only access
A predictable backend URL:
s3://pulumi-<repo>-state/state/<stack>

### Tasks
Include specific tasks in the order they need to be done. Include links to specific lines of code where the job should happen.
- [ ] Implement pulumi_state.py that creates the dedicated repo-specific S3 bucket.
- [ ] Add lifecycle, versioning, SSE-S3, TLS denial policy, and public access block.
- [ ] Export Pulumi backend URL from the module.
- [ ] Add github_oidc.py to create OIDC provider and per-repo deploy role.
- [ ] Implement least-privilege S3 policy allowing only state-object access.
- [ ] Wire modules into __main__.py.
- [ ] Test deployment in test AWS account.
- [ ] After validation, deploy the same code into the prod AWS account.

### Acceptance Criteria
Include specific criteria for how to prove that this issue will be done 
- [ ] Running pulumi up in test creates a bucket: pulumi-<repo>-state
- [ ] Bucket has: versioning, lifecycle, SSE-S3, TLS-only policy, public access block
- [ ] Deploy role PulumiDeploy-<repo> exists and matches repo/branch restrictions
- [ ] GitHub Actions can successfully run pulumi up using OIDC (no static keys)
- [ ] Backend URL works: pulumi config set backend-url s3://pulumi-…
- [ ] Same deployment succeeds in prod account with no code changes
- [ ] Documentation for onboarding new repos is included (config values required)

### PR DESCRIPTION
**Pull request by bra1n2h**
## Description
This PR implements repository-scoped Pulumi state infrastructure following AWS best practices.
It introduces automated provisioning of a dedicated S3 state bucket per repository, secure GitHub OIDC integration, and least-privilege IAM roles for Pulumi deployments.

## Related Issue
Resolves #1 

## Motivation and Context
This change establishes a secure, standardized, and automated pattern for managing Pulumi state across all repositories.
It removes the need for static AWS keys, enforces AWS security best practices, and ensures reproducible multi-account deployments.
New repositories can onboard by simply providing configuration values without modifying the core infra.

## How Has This Been Tested?
- Ran pulumi up in the test AWS account
  - Verified bucket creation: pulumi-<repo>-state
  - Confirmed versioning, lifecycle rules, SSE-S3, TLS-only access policy, and public access block
- Verified that exported backend URL matches expected format:
   s3://pulumi-<repo>-state/state/<stack>
- Validated GitHub Actions workflow using OIDC (no static credentials)
- Confirmed deploy role constraints match repo/branch restrictions
- Performed full deployment flow using test account → success
- Dry-run validation for prod account using same code (no changes required)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] I have performed a self-review of my code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING.md**](https://github.com/VilnaCRM-Org/infrastructure-template/blob/main/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] You have only one commit (if not, squash them into one commit).

## Summary by CodeRabbit

* **New Features**
  * Optional bootstrap provisioning via Pulumi config; centralized cross-region logging, automated S3 backups, per-repo Pulumi state buckets, secrets (KMS) and automation resources with GitHub Actions OIDC integration.
* **Tests**
  * Large suite of unit/integration tests added to cover config/sanitization, components, helpers, and project/CI contracts.

### CODE REVIEW

File: .env.empty
Comment (Kravalg) [inline .env.empty:1]: Please rename this file to the .env.example since this name is best practice in the python

File: .github/workflows/pulumi-integration.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.11"
+
+      - name: Install dependencies
+        run: |
+          curl -sSL https://install.python-poetry.org | python3 -
```
Comment (Kravalg) [inline .github/workflows/pulumi-integration.yml]: We should use docker for all GitHub CI checks

File: .github/workflows/pulumi.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
+name: pulumi
```
Comment (Kravalg) [inline .github/workflows/pulumi.yml]: This CI check should have understandable name, the pulumi isn't correct here
Please rename it

File: Makefile [hunk 3]
```diff
@@ -1,25 +1,84 @@
 # Parameters
-PROJECT        = infrastructure-template
-DOCKER_IMAGE   = ghcr.io/boltops-tools/terraspace:alpine
-ENV_FILE       = .env
+PROJECT            = bootstrap-infrastructure
+ENV_FILE           = .env
+EMPTY_ENV_FILE     = .env.empty
```
Comment (Kravalg) [inline Makefile:4]: Please remove all mentions of the .env.empty file and .env.example file and use .env committed file approach only for local development or CI checks
All other shared environment variables should be inside pulumi config

File: pulumi/__main__.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+"""
```
Comment (Kravalg) [inline pulumi/__main__.py]: The whole this file should be refactored
Please check how should look like the main.py file in the branch that I left

Main.py should call the class components that you should build instead of functions that you provided currently 

Here is the example

https://www.pulumi.com/docs/iac/guides/building-extending/components/build-a-component/

File: pulumi/infra/__init__.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+"""
```
Comment (Kravalg) [inline pulumi/infra/__init__.py]: Init py file shouldn't have such complicated logic and is just a facade that re-exports the public building blocks of your infra package

# pulumi/infra/__init__.py

```
from .config import BootstrapConfig
from .logging_bucket import LoggingBucket
from .pulumi_state import PulumiStateBackend
from .backup import BackupPlan

__all__ = [
    "BootstrapConfig",
    "LoggingBucket",
    "PulumiStateBackend",
    "BackupPlan",
]
```

File: pulumi/infra/backup.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
+import json
```
Comment (Kravalg) [inline pulumi/infra/backup.py:3]: All this files with the features should be refactored to the pulumi components as it was in the example that I've provided

**Review by Kravalg — DISMISSED** [review]
All python files with the DevOps code should refactored to the pulumi components and be calls from the main.py file as it was written previously by me in the example PR

**Comment by Kravalg** [comment]
@coderabbitai review\n\nUpdated PR to componentize Pulumi infra (logging/state/OIDC/backup), fixed OIDC branch binding, aligned githubToken config typing, tightened bucket existence error handling, and refreshed README onboarding.\n\nLocal tests: make test (unit passed; integration/structural skipped because directories missing).

**Comment by Kravalg** [comment]
@Kravalg addressed the componentization request: infra modules are now ComponentResources and instantiated from pulumi/__main__.py. Also fixed OIDC branch binding, githubToken typing, and backup/time windows. CI checks are green. Please re-review when you have a moment.

**Comment by Kravalg** [comment]
@coderabbitai review\n\nAddressed the remaining TRY300 suggestion by moving the success return into an else block in pulumi_state._bucket_exists. All prior feedback (OIDC branch binding, githubToken type, backup tags/windows, acl removal) is already fixed. Local tests: make test.

**Comment by Kravalg** [comment]
@coderabbitai review\n\nApplied security/AWS best-practice improvements:\n- Enforce TLS-only policy across full state buckets\n- Add abort-incomplete-multipart lifecycle (state + logging)\n- Enforce BucketOwnerEnforced ownership on state buckets (primary + replica)\n- Tag replication/IAM roles for traceability\n\nLocal tests: make test (unit passed; integration/structural skipped because dirs missing).

**Comment by Kravalg** [comment]
@coderabbitai review\n\nAdded missing docstrings to component __init__ methods to clear docstring coverage warning. Local tests: make test.

**Comment by Kravalg** [comment]
@coderabbitai review\n\nRe-triggering review after rate limit window.

**Comment by Kravalg** [comment]
@coderabbitai review\n\nRe-triggering review after rate limit window.

**Comment by dmytrocraft** [comment]
@coderabbitai review\n\nRe-triggering review.

**Comment by dmytrocraft** [comment]
@coderabbitai review\n\nRe-triggering review after hourly rate limit reset.

**Comment by Kravalg** [comment]
@coderabbitai full review

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,79 @@
+# syntax=docker/dockerfile:1.7-labs
+
+FROM python:3.11.9-slim-bookworm AS base
```
Comment (Kravalg) [inline Dockerfile]: Make sure that we are using as small as possible the docker base image (probably we can use alpine image)

**Review by Kravalg — CHANGES_REQUESTED** [review]
You need to refactor the PR regarding to the comments

**Comment by dmytrocraft** [comment]
Addressed all CodeRabbit actionable items and added AGENTS.md with the review-refactor workflow. Fixed Dockerfile Pulumi download (GitHub release checksums) and corrected bats-action tag; CI now green except CodeRabbit check failing due to rate limit. Please re-review when you have a moment.

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
Validated locally after the latest review fixes.

Changes in this update:
- replaced the brittle Poetry installer flow in the Docker image and trimmed build-only packages
- added a same-region replication guard and bounded IAM naming for Pulumi state replication
- fixed `_bucket_exists()` to handle live provider not-found responses during bootstrap
- updated stack smoke tests to use `monkeypatch` and added coverage for the new edge cases

Local validation:
- `docker compose build pulumi`
- `make test-unit`
- `make test-integration`
- live Pulumi smoke deployment against AWS account `891377212104` using a temporary S3 backend, followed by successful teardown to keep the account clean

@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai review

**Comment by dmytrocraft** [comment]
@coderabbitai full review

**Comment by dmytrocraft** [comment]
@coderabbitai resume

**Comment by dmytrocraft** [comment]
@coderabbitai full review

**Comment by dmytrocraft** [comment]
@coderabbitai full review

**Comment by dmytrocraft** [comment]
@coderabbitai full review

File: .env.empty
Comment (dmytrocraft) [reply .env.empty:1]: Addressed on the current head. The repo now uses a committed `.env` file for local/CI defaults, and the old `.env.empty` file is no longer present.

File: .github/workflows/pulumi-integration.yml (hunk 1 above)
Comment (dmytrocraft) [reply .github/workflows/pulumi-integration.yml]: Addressed on the current head. The integration workflow now builds the Pulumi Docker image and runs the test target through the containerized toolchain.

File: .github/workflows/pulumi.yml (hunk 2 above)
Comment (dmytrocraft) [reply .github/workflows/pulumi.yml]: Addressed on the current head. This workflow now has a specific name: `Pulumi Deploy Test`.

File: Makefile (hunk 3 above)
Comment (dmytrocraft) [reply Makefile:4]: Addressed on the current head. The Makefile now defaults to `.env`, and `.env.empty` / `.env.example` references were removed from the local workflow in favor of Pulumi config for shared settings.

File: pulumi/infra/__init__.py (hunk 5 above)
Comment (dmytrocraft) [reply pulumi/infra/__init__.py]: Addressed on the current head. `pulumi/infra/__init__.py` is now a simple facade that re-exports the public component classes only.

File: Dockerfile (hunk 7 above)
Comment (dmytrocraft) [reply Dockerfile]: Addressed on the current head. The Docker image was reworked into a multi-stage slim runtime and is now down to roughly 272 MB actual image size while keeping the Pulumi/AWS toolchain intact. Alpine was not used because AWS CLI v2 requires a glibc-based runtime.

File: Makefile [hunk 8]
```diff
@@ -1,25 +1,211 @@
...
+ENV_FILE           = .env
+COMPOSE_SERVICE   ?= pulumi
+PULUMI_DIR        ?= pulumi
+STACK             ?=
+PULUMI_SECRETS_PROVIDER ?=
+BATS_DOCKER_IMAGE ?= bats/bats:1.11.1
+EFFECTIVE_ENV_FILE := $(firstword $(wildcard $(ENV_FILE)))
+ACTIONLINT_IMAGE  ?= rhysd/actionlint:1.7.7
+CHECKOV_IMAGE     ?= bridgecrew/checkov:3.2.487
+HADOLINT_IMAGE    ?= hadolint/hadolint:latest
```
[the remark this answers is not in the document]
Comment (dmytrocraft) [reply Makefile]: Addressed on the current head. `HADOLINT_IMAGE` is now pinned to a concrete release (`hadolint/hadolint:v2.14.0`) instead of `latest`, so the workflow is reproducible.

File: .github/workflows/pulumi-pr-commands.yml [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+      - name: Checkout PR head
+        if: >-
+          steps.parse.outputs.skip != 'true' &&
+          steps.parse.outputs.authorized == 'true' &&
+          steps.pr.outputs.head_repo == github.repository
+        uses: actions/checkout@34e114876b0b11c390a56381ad16ebd13914f8d5 # v4
+        with:
+          ref: ${{ steps.pr.outputs.head_sha }}
+
+      - name: Configure AWS credentials
```
[the remark this answers is not in the document]
Comment (dmytrocraft) [reply .github/workflows/pulumi-pr-commands.yml]: Addressed on the current head. The privileged `issue_comment` workflow is now only a dispatcher and does not check out or execute PR code; execution moved to a separate trusted `repository_dispatch` runner with head-SHA validation.

[the remark this answers is not in the document]
Comment (dmytrocraft) [reply .github/workflows/pulumi-pr-commands.yml]: Addressed on the current head. The untrusted checkout path was removed from the `issue_comment` workflow; only the trusted runner workflow checks out the exact validated PR head SHA.

File: .github/workflows/autoprerelase.yml [hunk 10]
```diff
@@ -28,11 +32,10 @@
           pre-release: "true"
 
       - name: Create Release
-        uses: actions/create-release@v1
+        uses: softprops/action-gh-release@v2
```
[the remark this answers is not in the document]
Comment (dmytrocraft) [reply .github/workflows/autoprerelase.yml]: Addressed on the current head. `softprops/action-gh-release` is now pinned to a full commit SHA in this workflow.

File: .github/workflows/autorelease.yml [hunk 11]
```diff
@@ -27,11 +31,10 @@
           output-file: 'CHANGELOG.md'
 
       - name: Create Release
-        uses: actions/create-release@v1
+        uses: softprops/action-gh-release@v2
```
[the remark this answers is not in the document]
Comment (dmytrocraft) [reply .github/workflows/autorelease.yml]: Addressed on the current head. `softprops/action-gh-release` is now pinned to a full commit SHA in this workflow.

File: .github/workflows/pulumi-e2e.yml [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+      AWS_REGION: eu-central-1
+      PULUMI_E2E_SECRETS_PROVIDER: ${{ vars.PULUMI_TEST_SECRETS_PROVIDER }}
+    steps:
+      - uses: actions/checkout@34e114876b0b11c390a56381ad16ebd13914f8d5 # v4
+
+      - name: Build test container
+        run: docker compose build pulumi
+
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v4
```
[the remark this answers is not in the document]
Comment (dmytrocraft) [reply .github/workflows/pulumi-e2e.yml]: Addressed on the current head. `aws-actions/configure-aws-credentials` is pinned to a full commit SHA in this workflow.

File: .github/workflows/repo-health.yml [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    name: OpenSSF Scorecard
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@34e114876b0b11c390a56381ad16ebd13914f8d5 # v4
+        with:
+          persist-credentials: false
+
+      - name: Run Scorecard
+        uses: ossf/scorecard-action@v2.4.3
```
[the remark this answers is not in the document]
Comment (dmytrocraft) [reply .github/workflows/repo-health.yml]: Addressed on the current head. `ossf/scorecard-action` is pinned to a full commit SHA in this workflow.

**Comment by dmytrocraft** [comment]
@coderabbitai full review

All previously stale unresolved threads were replied to and resolved on the current head. Please re-check the PR.

**Review by Kravalg — DISMISSED** [review]
CodeRabbit should approve this pr

**Comment by dmytrocraft** [comment]
@coderabbitai full review

Current head is `9484435d79be532f05f0ccb218c99dd77877a505`. All actionable prior threads are resolved on the latest diff; please review the current head.

**Comment by dmytrocraft** [comment]
@coderabbitai review

Current head is `31737a3715edad5f7392bc90794f713510b55077`. The follow-up nits from the earlier `25a5c80` review are addressed and the current CodeRabbit check is green; please re-evaluate the latest head.

**Review by dmytrocraft — APPROVED** [review]
Validated the current head 31737a3715edad5f7392bc90794f713510b55077 after syncing with main, addressing the remaining CodeRabbit follow-ups, and rerunning focused local checks.

**Comment by dmytrocraft** [comment]
@coderabbitai review
